src/systems/filecoin_files/data_transfer/data_transfer_subsystem.id

```diff
@@ -1,66 +1,127 @@
-type Address struct {}
-type Storage struct {}
-type StorageDealID struct {}
-type Function struct {}
-type Piece struct {}
+import ipld "github.com/filecoin-project/specs/libraries/ipld"
+import libp2p "github.com/filecoin-project/specs/libraries/libp2p"
 
-type DataTransferSubsystem struct @(mutable) {
-    dataTransfers  {UInt: DataTransferChannel}
-    volumes        [Storage]
+import piece "github.com/filecoin-project/specs/systems/filecoin_files/piece"
+
+type StorageDeal struct {}
+type RetrievalDeal struct {}
+
+// A DataTransferVoucher is used to validate
+// a data transfer request against the underlying storage or retrieval deal
+// that precipitated it
+type DataTransferVoucher union {
+	StorageDealVoucher 
+	RetrievalDealVoucher
+}
+
+type StorageDealVoucher struct {
+	deal StorageDeal
+}
+
+type RetrievalDealVoucher struct {
+	deal RetrievalDeal 
+}
 
-    OpenDataTransferChannel(x StorageDealID) UInt
-    CloseDataTransferChannel(x UInt)
-    TransferChannelStatus(x UInt) DataTransferStatus
-    ReportDataTransferEvent(x Function) union {Ret DataTransferStatus, Err error}
+type Ongoing struct {}
+type Completed struct {}
+type Failed struct {}
+type ChannelNotFoundError struct {}
 
-    findData(x CID) Piece
+type DataTransferStatus union {
+	Ongoing
+	Completed
+	Failed
+	ChannelNotFoundError
 }
 
-// Assumes access to error-checked, ordered, reliable transmission protocol
-type Scheduler struct {
-    ScheduleTransfer(x UInt, y Piece)
-    requestData(x DataTransferChannel) union {Ret Bytes, Err error}
-    sendData(x DataTransferChannel, y Bytes) error?
+type TransferID UInt
+
+type ChannelID struct {
+	to libp2p.PeerID
+	id TransferID
```

Comment:
a bit confused:
- does that mean the requestor's `DataTransferSubsystem.dataTransfers` map has `ChannelIDs` where the `to` field is `DataTransferSubsystem.host.PeerID` ? (ie itself?).
  - if so, maybe `requesterPeer`
- or does `to` mean "a channel/connection to peerX"), and it contains "the other side's peerID" 
  - if so, maybe `peer` may be a clearer name, or `remote` or `remotePeer` (as libp2p calls it)
- or is `to` supposed to be "the responder" ("a requester sends a request **to** the responder")
  - if so maybe calling it `responder` or `responderPeer` makes sense?
- (my understanding is off -- please correct me. i'm trying to square the "identifier for this channel shared by request and responder, **set by requestor** through protocol" with one map on `ChannelID` only)
- the distinction above matters in cases where the same pair of peers can have two active `DataTransferChannels` with the same `TransferID` (can happen if both are requesting stuff from each other) - i think that's OK because the `TransferID` is relative to the `requester` peer, so it is always well defined. (even if both sides request stuff from each other, and collide in TransferID, it's still unambiguous. BUT im not sure if `ChannelID` is unambiguous, it depends on the above.
---
- i'm reminded that IP network stacks tend to keep 5-tuples `(proto, srcAddr, srcPort, dstAddr, dstPort)`  and that stream muxers like yamux/etc keep stream ids in each direction
  - maybe keeping a 3-tuple makes sense `type ChannelID { requesterPeer, responderPeer, requesterTransferID }` ? not sure
  - i think i agree that we dont need a 4-tuple: `type ChannelID { localPeer, remotePeer, localTransferID, remoteTransferID }` because one channel is not bidirectional, has a single direction.